hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/handler/AssignRegionHandler.java

```diff
@@ -105,50 +102,54 @@ public void process() throws IOException {
       }
       return;
     }
     LOG.info("Open {}", regionName);
     HRegion region;
+    // Start a try. Above we registered region with rs.getRegionsInTransitionInRS. Need to clear
+    // it on success or error on the end.
     try {
-      TableDescriptor htd =
-        tableDesc != null ? tableDesc : rs.getTableDescriptors().get(regionInfo.getTable());
-      if (htd == null) {
-        throw new IOException("Missing table descriptor for " + regionName);
+      try {
+        TableDescriptor htd =
+          tableDesc != null ? tableDesc : rs.getTableDescriptors().get(regionInfo.getTable());
+        if (htd == null) {
+          throw new IOException("Missing table descriptor for " + regionName);
+        }
+        // pass null for the last parameter, which used to be a CancelableProgressable, as now the
+        // opening can not be interrupted by a close request any more.
+        region = HRegion
+          .openHRegion(regionInfo, htd, rs.getWAL(regionInfo), rs.getConfiguration(), rs, null);
+      } catch (IOException e) {
+        cleanUpAndReportFailure(e);
+        return;
       }
-      // pass null for the last parameter, which used to be a CancelableProgressable, as now the
-      // opening can not be interrupted by a close request any more.
-      region = HRegion.openHRegion(regionInfo, htd, rs.getWAL(regionInfo), rs.getConfiguration(),
-        rs, null);
-    } catch (IOException e) {
-      cleanUpAndReportFailure(e);
-      return;
-    }
-    rs.postOpenDeployTasks(new PostOpenDeployContext(region, openProcId, masterSystemTime));
-    rs.addRegion(region);
-    LOG.info("Opened {}", regionName);
-    // Cache the open region procedure id after report region transition succeed.
-    rs.finishRegionProcedure(openProcId);
-    Boolean current = rs.getRegionsInTransitionInRS().remove(regionInfo.getEncodedNameAsBytes());
-    if (current == null) {
-      // Should NEVER happen, but let's be paranoid.
-      LOG.error("Bad state: we've just opened a region that was NOT in transition. Region={}",
-        regionName);
-    } else if (!current) {
-      // Should NEVER happen, but let's be paranoid.
-      LOG.error("Bad state: we've just opened a region that was closing. Region={}", regionName);
+      rs.postOpenDeployTasks(new PostOpenDeployContext(region, openProcId, masterSystemTime));
+      rs.addRegion(region);
+      LOG.info("Opened {}", regionName);
+    } finally {
+      // Remove from regionsInTransitionRS whether success or failure.
+      rs.getRegionsInTransitionInRS().remove(regionInfo.getEncodedNameAsBytes(), Boolean.TRUE);
+      Boolean current = rs.getRegionsInTransitionInRS().remove(regionInfo.getEncodedNameAsBytes());
+      if (current == null) {
+        // Should NEVER happen, but let's be paranoid.
+        LOG.error("Bad state: we've just opened {} and it was NOT in transition", regionName);
+      } else if (!current) {
+        // Should NEVER happen, but let's be paranoid.
```

Comment:
another lie, same thinking re: log level.

Reply:
This is not a lie, if this happens, we will be in trouble as there will be critical bugs in the code.

Reply:
Can you enlighten me mighty @apache9? By my read, any number of these handlers could be running concurrently, even on the same region. There's no mutex around access to the associated region's entry in the map, all we have is the promise of concurrent hashmap that concurrent edits won't tear. Thus, there's no guarantee that some change we made previously should be visible currently.